User can access entire access token response

jwoudenberg opened the pull request:
Hi there! Thank you for creating and maintaining this project, it's great!

I'm looking to put OAuth2 login in front of a WAI project. `wai-middleware-auth` is perfect for my use case, apart from one thing: Of the data returned by the OAuth2 provider in response to a token request and the end of the login flow only the access token is made available to the application. The response, described by the [`OAuth2Token`] type in `hoauth2` contains a number of other fields I'm interested in. I could retrieve them using the access token, but that would require an additional HTTP request to the OAuth2 provider request which seems wasteful.

I made some changes on a fork to support the personal project, giving access to the whole token response. Is there an interest in merging a change like that back?

If so I'm happy to clean up this PR a bit further in advance of a review. I wouldn't advice to merge these changes just yet, because:
- I need to test them a bit more thoroughly.
- They break the behavior of the existing `getAuthUser` function without breaking any types, so anyone upgrading to this version would be in for a nasty surprise at runtime.

I'm happy to fix these issues (and others that may come up in review), just wanted to briefly check if there's an interest for this change in the first place before putting in the effort.

Thank you for reading, and for your work on this library!

[`OAuth2Token`]: https://www.stackage.org/haddock/lts-14.8/hoauth2-1.8.9/Network-OAuth-OAuth2-Internal.html#t:OAuth2Token

jwoudenberg committed:
User can access entire access token response

As part of the Oauth2 login flow we get from the auth provider an access
token and potentially also a refresh token, id token, expiry date, and
token type. Of these `wai-middleware-auth` was only making the access
token available to the application behind the middleware, by way of the
`getAuthUser` function.

This commit gives access to the full response from the token provider
endpoint.

Not so nice is that the `getAuthUser` function now returns a
backwards-incompatible result for OAuth providers, without that being
explicit in the types. It might be nice to make an intentionally
breaking change to draw attention to this fact.

lehins commented:
@jwoudenberg This certainly makes sense. One recommendation I have is it would be better to re-use and re-export types from `hoauth2` instead of redefining similar ones (i.e. `OAuth2Token`). This will require a bump up in the minimum version in cabal file, but that is ok, since this way we could also get rid of the ugly CPP needed for backwards compatibility with `houath <= 1`

Changes you suggest will be backwards incompatible, since the binary format of the logged in user has changed, so next version will be `0.2.0` which is not a problem, but still important thing to note in the changelog.

lehins commented on src/Network/Wai/Middleware/Auth/OAuth2.hs:
```diff
@@ -180,3 +215,12 @@ instance AuthProvider OAuth2 where
+--
+-- If called on a @Request@ behind the middleware, should always return a
+-- @Just@ value.
+getAccessToken :: Request -> Maybe Token
```
Adding a `@since 0.2.0` will be needed here

jwoudenberg commented:
Thank you for the super quick response! Happy to hear this is a change you'd be interested in merging back.

> @jwoudenberg This certainly makes sense. One recommendation I have is it would be better to re-use and re-export types from hoauth2 instead of redefining similar ones (i.e. OAuth2Token). This will require a bump up in the minimum version in cabal file, but that is ok, since this way we could also get rid of the ugly CPP needed for backwards compatibility with houath <= 1

This sounds great! The backwards compatibility was the reason I didn't reuse the `OAuth2Token` type, the older version of `hoauth2` doesn't have it.

jwoudenberg committed:
Drop support for hoauth versions < 1.x

This means the code no longer compiles with stackage LTS 8.x, so we
upgrade that too.

jwoudenberg committed:
Reuse `hoauth2` access token type

The `OAuth2Token` type doesn't have a Binary instance, but it does have
`ToJSON`/`FromJSON` instances, so we use those for encoding/decoding.

jwoudenberg committed:
Bump version and update CHANGELOG.md

We need to do a major version bump because the encoded contents of the
oauth session will change. This means existing session cookies will not
be valid anymore.

jwoudenberg commented:
I made the changes you recommended. I see two remaining open questions (and maybe you see more):

- The `OAuth2Token` doesn't have a `Binary` instance so I used its `Aeson` instances to encode it. Does that seem fine? Or should I look into adding a `newtype` wrapper do allow encoding using a `Binary` instance  like before?
- The `AuthUser`'s `authUserIdentity` field now contains a JSON-encoded `OAuth2Token` instead of an access token. You've rightly pointed out this isn't a backwards-compatible change to begin with and so I'm less concerned about it then before, but an `OAuth2Token` is maybe not what people would expect as a 'user identity'. I'm not sure what's best here. I could add an additional field on the `AuthUser` to contain the full `OAuth2Token`, so `authUserIdentity` contains the access token alone like before. Any thoughts?

lehins commented:
I sort of expected this question: "The OAuth2Token doesn't have a Binary instance"

Using newtype would be the best way to go, I think, since you already added a helper function `getAccessToken` that does the decoding, I don't think it will be a problem, especially since `OAuth2Token` has a `Generic` instance already

Alternative would be to submit a PR to `hoauth` with such an instance (it already depends on `binary`, so it is unlikely that it wouldn't be merged) and declare an orphan instance for `Binary` in `wai-middleware-auth` with some CPP for houath2 version again :) This approach is the best one, I think, since it solves the problem at its root, but it is more involved and would take some time. So using a newtype wrapper is a good solution.

I'd be against using aeson instance for that purpose, it would unnecessarily increase the size of the cookie

The name `UserIdentity` is a bit misleading, I guess, I should have named it something like `AuthLoginState`. If you think it makes it clearer, I am ok with renaming it. That would entail a bit of deprecation:
```haskell
-- | An arbitrary state that comes with logged in user, eg. a username, token or an email address.
type AuthLoginState = S.ByteString`

type UserIdentity = S.ByteString
{-# DEPRECATED UserIdentity "In favor of `AuthLoginState`" #-}

authUserIdentity :: AuthUser -> UserIdentity
authUserIdentity = authLoginState
{-# DEPRECATED authUserIdentity' "In favor of `authLoginState`" #-}
```

jwoudenberg committed:
Encode OAuth2Token using Binary instance

jwoudenberg commented:
Yeah, I agree the `Binary` approach is much nicer.

> Using newtype would be the best way to go, I think, since you already added a helper function `getAccessToken` that does the decoding, I don't think it will be a problem, especially since `OAuth2Token` has a `Generic` instance already

I tried using the `Generic` instance available, but I ran aground on the nested `AccessToken`, `RefreshToken`, and `IdToken` types which don't have `Binary` instances themselves. I wrote custom encoder / decoder instead. The downside of that is that this strategy is less resilient to future changes in the `hoauth2` library, so even if it doesn't incur any CPP now it might in the future.

I think I'm going to make that PR against `hoauth2`, that seems to be a good investment regardless of which approach we take here. Let me know what you think of the custom decoder/encoder. If we move forward with that I think I should add a test to ensure `put` and `get` are duals.

It's getting late for me so I'm logging off for today, but I'll continue a bit tomorrow. Thank you for your support!

lehins commented on src/Network/Wai/Middleware/Auth/OAuth2.hs:
```diff
@@ -88,33 +84,28 @@ getClientSecret = id
+  OAuth2TokenBinary { unOAuth2TokenBinary :: OA2.OAuth2Token }
+
+instance Binary OAuth2TokenBinary where
+  put (OAuth2TokenBinary token) = do
```
Here is a much easier approach, just do what `binary` does when it derives it:
```
put (OAuth2TokenBinary token) = gput . from
get = OAuth2TokenBinary . to <$> gget
```

Haven't tried it, but I can't see why it wouldn't work

jwoudenberg replied:
I was convinced that would work too, but no such luck :). This is the error I get when I try it.

```
    • No instance for (Binary OA2.AccessToken)
        arising from a use of ‘gput’
```

lehins replied:
Oh yeah, that's right, `OAuth2Token` is full of newtype wrappers :)
It will be good to add those Binary instances upstream. Until then what you've added looks good.

jwoudenberg replied:
I created a PR to the `hoauth2`:

https://github.com/freizl/hoauth2/pull/106

I noticed that the `binary` package is a dependency of a demo executable of that package, but not the library itself, so there's a bit of a downside to adding these instances. We'll have the see how the maintainer responds.

lehins replied:
I see, I didn't realize that, since I only glanced at dependencies on hackage. Whatever the outcome of that above PR might be it is not gonna be a blocker for us. In either case good job on all of your PRs.

jwoudenberg committed:
Rename `authUserIdentity` to `authLoginState`

The `authUserIdentity` is a bit of a misnomer. It implies it contains
something like a user id but in reality it can be anything returned by
a login process, notably in the case of an oauth2 flow a number of
tokens.

jwoudenberg commented:
If we had a derived `Binary` instance it wouldn't worry me, but given the manual `Binary` instance I wrote I'm a bit concerned concerned about the `put` and `get` implementations potentially not being duals. Would you be opposed if I added a test suite containing to test that duality?

lehins commented:
@jwoudenberg I would be all in favor of a test suite. A QuickCheck property for testing the duality is a great idea. Feel free to add as many tests as you'd like.

jwoudenberg committed:
Add test for Binary encoding/decoding of tokens

jwoudenberg commented:
I pushed a test, let me know what you think of it! Unfortunately deriving a generator at the moment isn't feasible for the same reason that we cannot derive `Binary` instances, so I had to create a manual generator.

Unfortunately the test forces me to export the `OAuth2TokenBinary` type. I moved it to an `Internal` module to clarify it is not an intended part of the public API, but let me know if you prefer a different approach!

lehins commented:
Thank you for your work @jwoudenberg I'll make a release on hackage sometime this week, unless before that you'll think of some other improvements we could bundle in ;)

lehins commented:
@jwoudenberg In case you were waiting for it: https://hackage.haskell.org/package/wai-middleware-auth-0.2.0.0

jwoudenberg commented:
Thank you!
